hal/src/photon/concurrent_hal.cpp

```diff
@@ -114,11 +103,239 @@ os_result_t os_thread_join(os_thread_t* thread)
  * @param thread    The thread to clean up.
  * @return 0 on success.
  */
-os_result_t os_thread_cleanup(os_thread_t* thread)
+os_result_t os_thread_cleanup(os_thread_t thread)
 {
-    return wiced_rtos_delete_thread((wiced_thread_t*)thread);
+    vTaskDelete(thread);
+    return 0;
 }
 
+/**
+ * Map gthread handles to FreeRTOS task handles.
+ */
+typedef TaskHandle_t __gthread_t;
+
+
+class ThreadQueue
+{
+    QueueHandle_t queue;
+
+public:
+
+    ThreadQueue(UBaseType_t max_size) {
+        queue = xQueueCreate(max_size, sizeof(TaskHandle_t));
+    }
+
+    ~ThreadQueue() {
+        vQueueDelete(queue);
+    }
+
+
+    void enqueue(TicksType_t ticks=portMAX_DELAY) {
+        TaskHandle_t current = xTaskGetCurrentTaskHandle();
+        xQueueSend(queue, &current, ticks);
+    }
+
+    void sleep() {
+        enqueue();
+        vTaskSuspend(NULL);
+    }
+
+    bool wake() {
+        TaskHandle_t next;
+        if (xQueueReceive(queue, &next, 0)==pdTRUE) {
+            vTaskResume(next);
+            return true;
+        }
+        return false;
+    }
+
+    void wakeAll()
+    {
+        while (wake()) {}
+    }
+};
+
+
+class Lock
+{
+    SemaphoreHandle_t semaphore;
+
+public:
+
+    Lock() {
+        semaphore = xSemaphoreCreateMutex();
+    }
+
+    void acquire() {
+        while (xSemaphoreTake(semaphore, portMAX_DELAY)==pdFALSE);
+    }
+
+    void release() {
+        xSemaphoreGive(semaphore);
+    }
+
+    ~Lock() {
+        vSemaphoreDelete(semaphore);
+    }
+};
+
+
+class ConditionVariable
+{
+    ThreadQueue queue;
+
+    typedef std::unique_lock<std::mutex> lock_t;
+public:
+    ConditionVariable(UBaseType_t max_size) : queue(max_size) {}
+
+    void wait(lock_t* lock)
+    {
+        taskENTER_CRITICAL();
+        lock->unlock();
+        queue.enqueue();
+        taskEXIT_CRITICAL();
+        vTaskSuspend(NULL);
+        lock->lock();
+    }
```

Comment:
I think a deadlock can happen if the task is pre-empted between the call to taskEXIT_CRITICAL and vTaskSuspend, and then the other task signals the CV. vTaskResume will be called on the original task before vTaskSuspend.

Instead of using vTaskResume and vTaskSuspend, I think you'll need one [binary semaphore](http://www.freertos.org/xSemaphoreCreateBinary.html) per thread to handle the suspend/resume logic.